Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -1924,96 +2000,158 @@ public partial class ModuleResolver {
         ddVar = ddWhereConstraintsAre.Var;
         ddConstraint = ddWhereConstraintsAre.Constraint;
       }
-      List<ComprehensionExpr.BoundedPool> bounds;
-      bool constraintConsistsSolelyOfRangeConstraints;
-      if (ddVar == null) {
-        // There are no bounds at all
-        bounds = new List<ComprehensionExpr.BoundedPool>();
-        constraintConsistsSolelyOfRangeConstraints = true;
+
+      List<(NativeType, bool coversEntireIntegerRange)> bigEnoughNativeTypes;
+      var detectedRange = "";
+      if (dd.BaseType.NormalizeToAncestorType() is BitvectorType bitvectorAncestorType) {
+        bigEnoughNativeTypes = FigureOutNativeTypeForBitvectorNewtype(dd, bitvectorAncestorType, ddVar == null,
+          nativeTypeChoices ?? NativeTypes, nativeTypeChoices != null);
+      } else if (dd.BaseType.NormalizeToAncestorType() is IntType) {
+        bigEnoughNativeTypes = FigureOutNativeTypeForIntegerNewtype(dd, ddVar, ddConstraint,
+          nativeTypeChoices ?? NativeTypes, nativeTypeChoices != null, out detectedRange);
       } else {
-        bounds = DiscoverAllBounds_SingleVar(ddVar, ddConstraint, out constraintConsistsSolelyOfRangeConstraints);
+        // No native type available
+        bigEnoughNativeTypes = new List<(NativeType, bool coversEntireIntegerRange)>();
+      }
+      if (bigEnoughNativeTypes == null) {
+        // An error occurred while computing "bigEnoughNativeTypes"
+        return;
       }
 
-      // Find which among the allowable native types can hold "dd". Give an
-      // error for any user-specified native type that's not big enough.
-      var bigEnoughNativeTypes = new List<NativeType>();
-      BigInteger? lowBound = null;
-      BigInteger? highBound = null;
-      foreach (var bound in bounds) {
-        void UpdateBounds(BigInteger? lo, BigInteger? hi) {
-          if (lo != null && (lowBound == null || lowBound < lo)) {
-            lowBound = lo; // we found a more restrictive lower bound
-          }
-          if (hi != null && (highBound == null || hi < highBound)) {
-            highBound = hi; // we found a more restrictive lower bound
+      // Finally, of the big-enough native types, pick the first one that is supported by the selected target compiler.
+      foreach (var (nativeT, coversEntireIntegerRange) in bigEnoughNativeTypes) {
+        if (Options.Backend.SupportedNativeTypes.Contains(nativeT.Name)) {
```

Comment:
A reference from the resolve to the back-end, yuk!

Do the `NativeType` and `TargetTypeCoversAllBitPatterns` fields need to be set at this point? Can we not delay that until a later phase.